src/renderers/dom/client/MarkupMismatchError.js

```diff
@@ -0,0 +1,216 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule MarkupMismatchError
+ */
+
+'use strict';
+
+var ELEMENT_NODE_TYPE = 1;
+var TEXT_NODE_TYPE = 3;
+var COMMENT_NODE_TYPE = 8;
+
+// we cannot use class MarkupMismatchError extends Error {} because Babel cannot extend built-in
+// objects in a way that allows for instanceof checks.
+// this error subclassing code is copied and modified from
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#Custom_Error_Types
+// According to https://developer.mozilla.org/en-US/docs/MDN/About#Copyrights_and_licenses, all code in MDN
+// added on or after August 20, 2010 is public domain, and according to the wiki page history, the earliest
+// version of this code was added on June 7, 2011:
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error$revision/62425
+function MarkupMismatchError(message, node, serverVersion, clientVersion) {
+  this.name = 'MarkupMismatchError';
+  this.message = message || 'The pre-rendered markup did not match the component being rendered.';
+  this.node = node;
+  this.serverVersion = serverVersion;
+  this.clientVersion = clientVersion;
+  this.stack = (new Error()).stack;
+}
+MarkupMismatchError.prototype = Object.create(Error.prototype);
+MarkupMismatchError.prototype.constructor = MarkupMismatchError;
+
+/**
+ * throw an error when there is a child of the node in the client component tree
+ * that was not present in the server markup.
+ * @param {DOMElement} parent in the server markup
+ * @param {ReactElement} The React element of the child in the client component tree.
+ */
+function throwChildAddedError(node, child) {
+  let childDesc = 'An unknown type of child.';
+  if (typeof child === 'string') {
+    childDesc = `The text ${child}`;
+  } else if (child.type) {
+    childDesc = `A child of type <${child.type}>`;
```

Comment:
This will work for primitive DOM components (string type), but will print some ugly nastyness for composite components (function type).  There is some standard code/incantation to get the type/name of an element, should find it and apply it here.